### ISSUE TITLE
Can we generalize the distribution logic for different record value types?

### ISSUE THREAD

**Issue by korthout** [camunda/zeebe#11218]
**Description**

Presently, only 1 record value type needs to be distributed over the partitions: Deployments. However, we are currently researching how to distribute other record value types: [Signals](https://github.com/camunda/zeebe/issues/10777) and [Deployment Deletions](https://github.com/camunda/zeebe/issues/9576). There may even be more targets for distributions in the future (e.g. process / process instance / job suspension/re-activation). 

We should investigate whether we can generalize the distribution logic. Generalization would allow us:
- to avoid building the same distribution concept multiple times for different parts of the workflow engine
- to avoid having to maintain multiple duplications of the same distribution concept (i.e. if we want to make changes to the distribution concept, we'd have to do it multiple times)

This would require us to:
- deprecate the existing records for Deployment distribution
- explore a generalized way to distribute arbitrary records

**Proposal**
Introduce a new value type called `RecordDistribution`, which wraps the record we want to communicate reliably to other partitions.

An initial `STARTED` intent ensures we store that record in the state (for lookups at retry), and a final `FINISHED` event can be applied to remove this record from the state.

Then, for every partition, we have intents `DISTRIBUTING`, `ACKKNOWLEDGE`, and `ACKNOWLEDGED` to keep track of the distribution to that specific partition.

On the receiving partition, we write the encapsulated record directly. Its key contains the information that it was distributed (and from which partition), so the engine knows the send the `ACKNOWLEDGE` back to the distributing partition.

**Comment by korthout** [comment]
We want to work on this rather sooner than later, but don't plan to work on it in the next iteration (next 2 weeks).

**Comment by npepinpe** [comment]
By distribution, do you mean broadcast to all partitions? If not, are message records not distributed as well?

I would also add, it would be great if timer start events were distributed evenly too :eyes:

**Comment by korthout** [comment]
By distribution, we mean the mechanism to broadcast an entity over all partitions **reliably**:
- send using [inter-partition command sender](https://github.com/camunda/zeebe/blob/main/stream-platform/src/main/java/io/camunda/zeebe/stream/api/InterPartitionCommandSender.java)
- send an acknowledgement from the receiving partition
- process the acknowledgment to stop retrying
- retry periodically if no acknowledgement has been received

>If not, are message records not distributed as well?

Message records are not distributed to **all** partitions.

>it would be great if timer start events were distributed evenly too 

Sure, but that's out of scope for this.

**Comment by npepinpe** [comment]
> Message records are not distributed to all partitions.

I'm aware, which is why I said "If not" :sweat_smile: Everything I wrote after "do you mean broadcast to all partitions" was contingent on distribution **not** meaning broadcast, but I see that's not the case. So yeah, timer start events would definitely be out of scope :+1: 

Thanks for clarifying!

**Comment by korthout** [comment]
Here's what it could look like for the distribution of the Deployment Create command

![deployment_record_distribution](https://user-images.githubusercontent.com/3511026/210559012-a4bfd048-ffc8-4925-9c61-1b24499a3e10.png)

And this would also work for the distribution of the Signal Broadcast command

![signal_record_distribution](https://user-images.githubusercontent.com/3511026/210559160-2c3c3d4f-e064-426e-8631-5d6980cde41f.png)

We can easily that the record distribution logic can be generalized.

> **Note** that I've pulled the Signal Broadcasted event forward (before distribution), but this is not necessary.

> **Note** that the signal logic still needs to have awareness about whether or not the signal broadcast command was distributed from other partitions or not.

![signal_extracted_distribution](https://user-images.githubusercontent.com/3511026/210559315-4bdb64b2-6b43-4277-8b2d-c3b9524e95c0.png)

Looking at this, I'm not sure whether Distribution is still the right word for it. Perhaps it is, but I'm open to new ideas.

**Comment by korthout** [comment]
One open question remains: how to deal with idempotence?

If the record under distribution can be retried and is successfully written to the other partition, but the other partition has no way to acknowledge the record distribution, then the receiving partition will have the same record distributed to it multiple times. Processing the same command multiple times should be done idempotently.

Currently, this problem does not matter for Deployments. We simply write the follow-up events once more and overwrite the state (i.e. using `upsert` instead of `insert`). Since the state doesn't really change, it's all good. But we cannot do that for signals. 

I've previously written about [dealing with signal distribution idempotently](https://github.com/camunda/zeebe/issues/10851#issuecomment-1335280337). That concept is specific to signals, and we could use it for signals explicitly. But something similar will have to be done for other distributions like process/job pausing/resuming. We should consider whether a generalized way of dealing with the idempotence of record distribution is possible.

**Comment by korthout** [comment]
@Zelldon Let's keep this engine specific for now

**Comment by korthout** [comment]
Until today, the biggest issue I saw was how to store a different record's value inside the `RecordDistributionRecord`.

I thought to remember that we did this for `Incidents` back before strict event-sourcing was introduced with v1. When I looked into it, however, this was not the case. The record under processing that results in the incident was wrapped in an `IndexedRecord` and stored in the state but not on the log. Back then, this was allowed. So I could not reuse any existing code because we haven't done this before.

However, I was inspired by [IndexedRecord's way of writing the process instance record](https://github.com/camunda/zeebe/blob/clients/go/v8.1.6/engine/src/main/java/io/camunda/zeebe/engine/state/instance/IndexedRecord.java#L68-L76) into a persistable form. This I could use to set any record value as the property of another by storing it as an `UnifiedRecordValue`.

```java
private final ObjectProperty<UnifiedRecordValue> recordValueProperty =
  new ObjectProperty<>("recordValue", new UnifiedRecordValue());
```

```java
public RecordDistributionRecord setRecordValue(final UnifiedRecordValue recordValue) {
  final var valueBuffer = new UnsafeBuffer(0, 0);
  final int encodedLength = recordValue.getLength();
  valueBuffer.wrap(new byte[encodedLength]);

  recordValue.write(valueBuffer, 0);
  recordValueProperty.getValue().read(new MsgPackReader().wrap(valueBuffer, 0, encodedLength));

  return this;
}
```

Of course, another problem became clear when trying to implement the `getRecordValue()` method for the `RecordDistributionRecord`. If you get an `UnifiedRecordValue` out of the property, it's no longer the same concrete `RecordValue` as it was written. This is a typical polymorphic deserialization problem we see more often in Java, which we have already solved in the `CommandApiRequestReader`. So using the same tactic of storing individual concrete record values in a map and fetching the right one based on the stored value type, we have a concrete record value to fill with data. All that was left was just a bit of writing bytes between buffers.

```java
static final Map<ValueType, Supplier<UnifiedRecordValue>> RECORDS_BY_TYPE =
    new EnumMap<>(ValueType.class);

// You'll need to register any of the record value's that you want to distribute
static {
  RECORDS_BY_TYPE.put(ValueType.DEPLOYMENT, DeploymentRecord::new);
}
```

```java
@Override
public RecordValue getRecordValue() {
  // fetch a concrete instance of the record value by type
  if (!valueTypeProperty.hasValue()) {
    throw new MsgpackException("Expected to read the value type property, but it's not yet set");
  }
  final var valueType = getValueType();
  final var concreteRecordValue = RECORDS_BY_TYPE.get(valueType).get();
  if (concreteRecordValue == null) {
    throw new IllegalStateException(
        "Expected to read the record value, but it's type `"
            + valueType.name()
            + "` is unknown. Please add it to RecordDistributionRecord.RECORDS_BY_TYPE");
  }

  // write the record value property's content into a buffer
  final var storedRecordValue = recordValueProperty.getValue();
  final var recordValueBuffer = new UnsafeBuffer(0, 0);
  final int encodedLength = storedRecordValue.getEncodedLength();
  recordValueBuffer.wrap(new byte[encodedLength]);
  storedRecordValue.write(new MsgPackWriter().wrap(recordValueBuffer, 0));

  // read the value back from the buffer into the concrete record value
  concreteRecordValue.wrap(recordValueBuffer);
  return concreteRecordValue;
}
```

You can find an implementation of this spike in [korthout-11218-spike](https://github.com/camunda/zeebe/compare/korthout-11218-spike). It also contains tests in [JsonSerializableToJsonTest.java](https://github.com/camunda/zeebe/compare/korthout-11218-spike#diff-ae4f86014bc41184dadc024691c43a2327a86e11de7f732d65839161145d8f79) to show that it works.

I wonder if there is more to prototype before we just implement the refactoring. So I'm calling this a success.

The next step is for the team to decide whether we want to take this approach, which we can discuss next week.

**Comment by korthout** [comment]
Yesterday, the team discussed the proposal described in this issue. The team likes the idea and we want to move forward with this.

**TL;DR**:
- We move ahead with the solution
- We discuss with ZDP team about doubling the current `maxMessageSize`
- We discuss with ZDP team about de-deprecating [a method in the InterPartitionCommandSender](https://github.com/camunda/zeebe/blob/8.2.0-alpha3/stream-platform/src/main/java/io/camunda/zeebe/stream/api/InterPartitionCommandSender.java#L32)
- After these discussions, we will breakdown the topic into tasks and adjust affected topics' task breakdowns
- We will hide difference in logic between distributed and undistributed commands behind an interface that we will find during a mob programming session
- We'll move most of the code for distribution in the engine close to the boundary with the streaming platform

**One main problem was raised**. For deployment, this solution requires us to write the deployed resources twice in the same record batch, i.e. once in the `Process:Created` (or `DecisionRequirements:Created`) events, and once in the `RecordDistribution:Started` event. That means, that we reach the `maxMessageSize` twice as fast. In other words, users will no longer be able to deploy some of the resources they could before. **The team proposes to solve this by doubling the current `maxMessageSize`**. We will discuss this with the Zeebe Distributed Platform team, as they may challenge this.

Another topic that should be discussed with the Zeebe Distributed Platform team is de-deprecating the method to [send a command to another partition where the sending partition determines the record key](https://github.com/camunda/zeebe/blob/8.2.0-alpha3/stream-platform/src/main/java/io/camunda/zeebe/stream/api/InterPartitionCommandSender.java#L32). This method was deprecated immediately when it was introduced with the new InterPartitionCommandSender API to avoid us from using it for other cases than Deployment Distribution. However, we could not figure out the reason why this method should not be used. Actually, the team sees reasonable use cases where a partition can decide the record key when it sends it to another partition. As the key encodes the partition id that it was generated on, it clearly identities an entity to exist on that specific partition. We should discuss with the Zeebe Distributed Platform team to understand why this method was deprecated in the first place, and push to revise that stance.

The team discussed the alternative between directly sending the command to distribute to the other partition versus introducing a special `DISTRIBUTE` intent for these records. The argument is that different logic may be required depending on whether the command is received initially, or distributed to that partition. The `DISTRIBUTE` intent nicely separates that logic, but leaks part of the distribution responsibility into each value type that should be distributed. Instead, the team settled on following the proposals idea of writing the command as is, and hiding the difference in logic behind a nice interface. We can look for this interface in an upcoming mob programming session to share more knowledge about the deployment distribution code.

Lastly, the team discussed how to send the `ACKNOWLEDGE` command to the other partition. This should happen in a post commit task, which we want to schedule outsie of the actual command processor of that specific value type, but in the Engine. By doing this close to the boundary between the platform and the engine, we may be able to move the distribution logic over into the streaming platform if this is ever needed.

Once the two points above have been discussed with the Zeebe Distributed Platform Team, we can breakdown the topic into tasks, and start implementation in parallel to adjusting the task breakdowns of the affected topics (i.e. deployment deletion, signals, etc).

**Comment by korthout** [comment]
I'm closing this issue, as the path forward is clear. The open questions will be discussed in 
- #11456
